pkg/cmd/project/link/link.go

```diff
@@ -0,0 +1,195 @@
+package link
+
+import (
+	"fmt"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/v2/api"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/pkg/cmd/project/shared/client"
+	"github.com/cli/cli/v2/pkg/cmd/project/shared/queries"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/shurcooL/githubv4"
+	"github.com/spf13/cobra"
+	"net/http"
+	"strconv"
+)
+
+type linkOpts struct {
+	number    int32
+	owner     string
+	repo      string
+	team      string
+	projectID string
+	repoID    string
+	teamID    string
+	format    string
+	exporter  cmdutil.Exporter
+}
+
+type linkConfig struct {
+	httpClient func() (*http.Client, error)
+	client     *queries.Client
+	opts       linkOpts
+	io         *iostreams.IOStreams
+}
+
+type linkProjectToRepoMutation struct {
+	LinkProjectV2ToRepository struct {
+		Repository queries.Repository `graphql:"repository"`
+	} `graphql:"linkProjectV2ToRepository(input:$input)"`
+}
+
+type linkProjectToTeamMutation struct {
+	LinkProjectV2ToTeam struct {
+		Team queries.Team `graphql:"team"`
+	} `graphql:"linkProjectV2ToTeam(input:$input)"`
+}
+
+func NewCmdLink(f *cmdutil.Factory, runF func(config linkConfig) error) *cobra.Command {
+	opts := linkOpts{}
+	linkCmd := &cobra.Command{
+		Short: "Link a project to a repository or a team",
+		Use:   "link [<number>] [flag]",
+		Example: heredoc.Doc(`
+			# link monalisa's project 1 to her repository "my_repo"
+			gh project link 1 --owner monalisa --repo my_repo
+
+			# link monalisa's organization's project 1 to her team "my_team"
+			gh project link 1 --owner my_organization --team my_team
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := client.New(f)
+			if err != nil {
+				return err
+			}
+
+			if len(args) == 1 {
+				num, err := strconv.ParseInt(args[0], 10, 32)
+				if err != nil {
+					return cmdutil.FlagErrorf("invalid number: %v", args[0])
+				}
+				opts.number = int32(num)
+			}
+
+			config := linkConfig{
+				httpClient: f.HttpClient,
+				client:     client,
+				opts:       opts,
+				io:         f.IOStreams,
+			}
+
+			if config.opts.repo != "" && config.opts.team != "" {
+				return fmt.Errorf("specify only one repo or team")
+			} else if config.opts.repo == "" && config.opts.team == "" {
+				return fmt.Errorf("specify at least one repo or team")
+			}
+
+			// allow testing of the command without actually running it
+			if runF != nil {
+				return runF(config)
+			}
+			return runLink(config)
+		},
+	}
+
+	linkCmd.Flags().StringVar(&opts.owner, "owner", "", "Login of the owner. Use \"@me\" for the current user.")
+	linkCmd.Flags().StringVarP(&opts.repo, "repo", "R", "", "The repository to be linked to this project")
+	linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project")
```

Comment:
Generally want to default to lowercase flags to avoid of extra keystrokes but also `-R,--repo  [HOST/]OWNER/REPO` is a standard flag on nearly every other command.

```suggestion
	linkCmd.Flags().StringVarP(&opts.repo, "repo", "r", "", "The repository to be linked to this project")
	linkCmd.Flags().StringVarP(&opts.team, "team", "t", "", "The team to be linked to this project")
```

Reply:
@andyfeller There is already a shorthand `t` for `--template`😱

Reply:
Ack, I went through this discovery as well, `-R` and `-T` are fine for me.

Reply:
linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project
```